File: android-client/app/src/main/java/com/componentvault/android/data/LcscCatalogRoutePolicy.kt
package com.componentvault.android.data

import com.componentvault.android.model.AppLanguage
import java.io.IOException
import java.util.Locale

internal enum class LcscCatalogSource(val wireName: String) {
    Domestic("lcsc_domestic_web"),
    International("lcsc_public_web"),
}

internal enum class LcscCatalogFailureKind {
    Blocked,
    RateLimited,
    CoolingDown,
    Unreachable,
    NoMatch,
    InvalidResponse,
}

internal enum class LcscDomesticGateReason { Blocked, Network, RateLimited }
internal class LcscDomesticCoolingDownException(val reason: LcscDomesticGateReason) : IOException("LCSC domestic lookup cooling down")

internal class LcscDomesticCooldownGate(private val now: () -> Long = System::currentTimeMillis) {
    private var until = 0L
    private var reason: LcscDomesticGateReason? = null
    @Synchronized fun current(): LcscDomesticGateReason? = reason?.takeIf { now() < until }.also { if (it == null) { reason = null; until = 0 } }
    @Synchronized fun record(value: LcscDomesticGateReason, retryAfterSeconds: Long? = null) {
        val seconds = when (value) {
            LcscDomesticGateReason.Blocked -> 120L
            LcscDomesticGateReason.Network -> 30L
            LcscDomesticGateReason.RateLimited -> (retryAfterSeconds ?: 30L).coerceIn(1L, 600L)
        }
        reason = value; until = now() + seconds * 1000L
    }
    @Synchronized fun clear() { reason = null; until = 0 }
}
internal val LcscDomesticSessionGate = LcscDomesticCooldownGate()

internal sealed interface LcscDomesticHttpFailure {
    data object Blocked : LcscDomesticHttpFailure
    data class RateLimited(val retryAfterSeconds: Long?) : LcscDomesticHttpFailure
    data class Http(val status: Int) : LcscDomesticHttpFailure
}

internal fun classifyDomesticHttpResponse(status: Int, body: String, retryAfter: String?): LcscDomesticHttpFailure? {
    if (status in 200..299) return null
    val challenge = status in setOf(403, 429, 503) &&
        listOf("_xvasu", "_xvtsc", "_xvpfs", "_xvpts", "Security verification")
            .any { body.contains(it, ignoreCase = true) }
    if (challenge) return LcscDomesticHttpFailure.Blocked
    if (status == 429) return LcscDomesticHttpFailure.RateLimited(retryAfter?.toLongOrNull()?.coerceIn(1, 600))
    return LcscDomesticHttpFailure.Http(status)
}

internal data class LcscCatalogAttempt(
    val source: LcscCatalogSource,
    val failure: LcscCatalogFailureKind,
)

internal data class LcscCatalogLookupResult(
    val metadata: com.componentvault.android.model.ComponentOfficialMetadata?,
    val preferredSource: LcscCatalogSource,
    val resolvedSource: LcscCatalogSource?,
    val attempts: List<LcscCatalogAttempt>,
) {
    val usedFallback: Boolean get() = resolvedSource != null && resolvedSource != preferredSource
}

internal object LcscCatalogRoutePolicy {
    fun preferredSource(language: AppLanguage): LcscCatalogSource = when (language) {
        AppLanguage.ZhCn -> LcscCatalogSource.Domestic
        AppLanguage.English -> LcscCatalogSource.International
    }

    fun preferredSource(locale: Locale): LcscCatalogSource =
        if (locale.language.equals("zh", ignoreCase = true)) {
            LcscCatalogSource.Domestic
        } else {
            LcscCatalogSource.International
        }

    fun order(preferred: LcscCatalogSource): List<LcscCatalogSource> = when (preferred) {
        LcscCatalogSource.Domestic -> listOf(LcscCatalogSource.Domestic, LcscCatalogSource.International)
        LcscCatalogSource.International -> listOf(LcscCatalogSource.International, LcscCatalogSource.Domestic)
    }

    fun shouldPersist(result: LcscCatalogLookupResult): Boolean =
        result.metadata != null && result.resolvedSource == result.preferredSource

    fun classify(error: Throwable): LcscCatalogFailureKind = when (error) {
        is LcscDomesticBlockedException -> LcscCatalogFailureKind.Blocked
        is LcscDomesticRateLimitedException -> LcscCatalogFailureKind.RateLimited
        is LcscDomesticCoolingDownException -> LcscCatalogFailureKind.CoolingDown
        is LcscDomesticResponseException -> LcscCatalogFailureKind.InvalidResponse
        is IOException -> LcscCatalogFailureKind.Unreachable
        else -> LcscCatalogFailureKind.InvalidResponse
    }
}

File: windows-client/ComponentVault.WinUI/Services/Catalog/LcscCatalogRoutePolicy.cs
using System.Globalization;

namespace ComponentVault.WinUI.Services.Catalog;

public enum LcscCatalogSource { Domestic, International }
public enum LcscCatalogFailureKind { Blocked, Unreachable, NoMatch, InvalidResponse }

public sealed record LcscCatalogAttempt(LcscCatalogSource Source, LcscCatalogFailureKind Failure);

public sealed record LcscCatalogLookupResult(
    LcscProductMetadata? Metadata,
    LcscCatalogSource PreferredSource,
    LcscCatalogSource? ResolvedSource,
    IReadOnlyList<LcscCatalogAttempt> Attempts)
{
    public bool UsedFallback => ResolvedSource is not null && ResolvedSource != PreferredSource;

    public string UserMessage
    {
        get
        {
            var english = PreferredSource == LcscCatalogSource.International;
            string Source(LcscCatalogSource source) => source == LcscCatalogSource.Domestic
                ? (english ? "LCSC China" : "国内立创") : (english ? "international LCSC" : "国际 LCSC");
            string Failure(LcscCatalogFailureKind failure) => failure switch
            {
                LcscCatalogFailureKind.Blocked => english ? "site verification blocked the request" : "站点验证阻断",
                LcscCatalogFailureKind.Unreachable => english ? "network or site unavailable" : "网络或站点不可达",
                LcscCatalogFailureKind.NoMatch => english ? "no exact match" : "未找到精确匹配",
                _ => english ? "unexpected site response" : "站点响应格式异常",
            };
            if (ResolvedSource is { } resolved && !UsedFallback)
                return english ? $"Exact metadata loaded from {Source(resolved)}." : $"已从{Source(resolved)}取得精确资料。";
            if (ResolvedSource is { } fallback)
                return english
                    ? $"Preferred source {Source(PreferredSource)} failed ({Failure(Attempts[0].Failure)}); using {Source(fallback)}."
                    : $"首选{Source(PreferredSource)}查询失败（{Failure(Attempts[0].Failure)}），已改用{Source(fallback)}。";
            return string.Join(english ? "; " : "；", Attempts.Select(x => $"{Source(x.Source)}{(english ? ": " : "：")}{Failure(x.Failure)}"));
        }
    }
}

public sealed record LcscCatalogSearchResult(
    IReadOnlyList<LcscProductMetadata> Items,
    LcscCatalogSource Source,
    string? Notice);

public static class LcscCatalogRoutePolicy
{
    public static LcscCatalogSource PreferredSource(CultureInfo culture) =>
        culture.TwoLetterISOLanguageName.Equals("zh", StringComparison.OrdinalIgnoreCase)
            ? LcscCatalogSource.Domestic : LcscCatalogSource.International;

    public static IReadOnlyList<LcscCatalogSource> Order(LcscCatalogSource preferred) =>
        preferred == LcscCatalogSource.Domestic
            ? [LcscCatalogSource.Domestic, LcscCatalogSource.International]
            : [LcscCatalogSource.International, LcscCatalogSource.Domestic];

    public static LcscCatalogFailureKind Classify(Exception exception) => exception switch
    {
        LcscDomesticBlockedException => LcscCatalogFailureKind.Blocked,
        InvalidDataException => LcscCatalogFailureKind.InvalidResponse,
        HttpRequestException or TaskCanceledException => LcscCatalogFailureKind.Unreachable,
        _ => LcscCatalogFailureKind.InvalidResponse,
    };
}
